cluster-autoscaler/utils/scheduler/scheduler.go

```diff
@@ -106,3 +117,31 @@ func ResourceToResourceList(r *schedulerframework.Resource) apiv1.ResourceList {
 	}
 	return result
 }
+
+// GetSchedulerConfig loads scheduler config from a path.
+func GetSchedulerConfig(path string) (*scheduler_config.KubeSchedulerConfiguration, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", SchedulerConfigLoadErr, err)
+	}
+
+	obj, gvk, err := scheduler_scheme.Codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", SchedulerConfigDecodeErr, err)
+	}
+
+	cfgObj, ok := obj.(*scheduler_config.KubeSchedulerConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("%s, gvk: %s", SchedulerConfigTypeCastErr, gvk)
+	}
+
+	// this needs to be set explicitly because config's api version is empty after decoding
+	// check kubernetes/cmd/kube-scheduler/app/options/configfile.go for more info
+	cfgObj.TypeMeta.APIVersion = gvk.GroupVersion().String()
```

Comment:
This function is based on upstream scheduler's private function called `loadConfigFromFile`

```go
File: cmd/kube-scheduler/app/options/configfile.go
34: func loadConfigFromFile(logger klog.Logger, file string) (*config.KubeSchedulerConfiguration, error) {
35: 	data, err := os.ReadFile(file)
36: 	if err != nil {
37: 		return nil, err
38: 	}
39: 
40: 	return loadConfig(logger, data)
41: }
42: 
43: func loadConfig(logger klog.Logger, data []byte) (*config.KubeSchedulerConfiguration, error) {
44: 	// The UniversalDecoder runs defaulting and returns the internal type by default.
45: 	obj, gvk, err := scheme.Codecs.UniversalDecoder().Decode(data, nil, nil)
46: 	if err != nil {
47: 		return nil, err
48: 	}
49: 	if cfgObj, ok := obj.(*config.KubeSchedulerConfiguration); ok {
50: 		// We don't set this field in pkg/scheduler/apis/config/{version}/conversion.go
51: 		// because the field will be cleared later by API machinery during
52: 		// conversion. See KubeSchedulerConfiguration internal type definition for
53: 		// more details.
54: 		cfgObj.TypeMeta.APIVersion = gvk.GroupVersion().String()
55: 		switch cfgObj.TypeMeta.APIVersion {
56: 		case configv1beta2.SchemeGroupVersion.String():
57: 			logger.Info("KubeSchedulerConfiguration v1beta2 is deprecated in v1.25, will be removed in v1.28")
58: 		case configv1beta3.SchemeGroupVersion.String():
59: 			logger.Info("KubeSchedulerConfiguration v1beta3 is deprecated in v1.26, will be removed in v1.29")
60: 		}
61: 		return cfgObj, nil
62: 	}
63: 	return nil, fmt.Errorf("couldn't decode as KubeSchedulerConfiguration, got %s: ", gvk)
64: }


```
https://github.com/kubernetes/kubernetes/blob/6b34fafdaf5998039c7e01fa33920a641b216d3e/cmd/kube-scheduler/app/options/configfile.go#L34-L64

Note that I haven't included the deprecated log messages because they might change and we'd have to remember updating them every time. I don't see a lot of value in printing these messages because we'd have to support the scheduler config even if it is deprecated anyway (open to other opinions here).

Note that if the user uses an old scheduler api version which isn't supported anymore like say `v1alpha1`, `UniversalDecoder().Decode` throws an error like this:
```
panic: no kind "KubeSchedulerConfiguration" is registered for version "kubescheduler.config.k8s.io/v1alpha1" in scheme "pkg/runtime/scheme.go:100"
```

Reply:
Thanks for the comment. I thought about this but I was also on the fence because scheduler has a similar function to write the config to a file
```go
File: cmd/kube-scheduler/app/options/configfile.go
091: // LogOrWriteConfig logs the completed component config and writes it into the given file name as YAML, if either is enabled
092: func LogOrWriteConfig(logger klog.Logger, fileName string, cfg *config.KubeSchedulerConfiguration, completedProfiles []config.KubeSchedulerProfile) error {
093: 	loggerV := logger.V(2)
094: 	if !loggerV.Enabled() && len(fileName) == 0 {
095: 		return nil
096: 	}
097: 	cfg.Profiles = completedProfiles
098: 
099: 	buf, err := encodeConfig(cfg)
100: 	if err != nil {
101: 		return err
102: 	}
103: 
104: 	if loggerV.Enabled() {
105: 		loggerV.Info("Using component config", "config", buf.String())
106: 	}
107: 
108: 	if len(fileName) > 0 {
109: 		configFile, err := os.Create(fileName)
110: 		if err != nil {
111: 			return err
112: 		}
113: 		defer configFile.Close()
114: 		if _, err := io.Copy(configFile, buf); err != nil {
115: 			return err
116: 		}
117: 		logger.Info("Wrote configuration", "file", fileName)
118: 		os.Exit(0)
119: 	}
120: 	return nil
121: }

```
https://github.com/kubernetes/kubernetes/blob/6b34fafdaf5998039c7e01fa33920a641b216d3e/cmd/kube-scheduler/app/options/configfile.go#L91-L121

Notice how it is calling `os.Exit(0)` on line 118. If you import and use this function in your code, the program is going to exit. My concern is, if we don't implement the function to read config file on our own we might make ourselves dependent on the implementation of `LoadConfigFromFile` (if `loadConfigFromFile` is converted to a public function) whose implementation is dependent on use-case of scheduler and not CA. If there's a change in the `LoadConfigFromFile` in the future, it will break things for us. That being said, same thing can be argued for `Filter` and `PostFilter` extension points but I think there's an important difference between `Filter`/`PostFilter` extension points and `LoadConfigFromFile` or `LogOrWriteConfig`. Former is similar to a public API and it is unlikely to change without a strong deprecation policy or notice to users but that might not be the case with latter. 

There's other options here as well:
Option 1:
* Get this PR in to CA codebase
* Talk with sig-scheduling about setting expectation about `LoadConfigFromFile` 
* Create another PR to replace our implementation with sig-scheduling's implementation (might have to wait until they release 1.28 so that we can import their code or ask for a patch to 1.27)

Option 2:
* Put this PR on hold
* Talk with sig-scheduling about setting expectation about `LoadConfigFromFile` 
* Create another PR to replace our implementation with sig-scheduling's implementation (might have to wait until they release 1.28 so that we can import their code or ask for a patch to 1.27)

If we decide to go with asking for a public implementation of `LoadConfigFromFile` and we want to go with Option 1, I will create a separate issue to co-ordinate with sig-scheduling and use the public implementation. If we go with Option 2, we will have to put this PR on hold until we get `LoadConfigFromFile`. 

Let me know what you think.

Reply:
PR to update scheduler code: https://github.com/kubernetes/kubernetes/pull/119057

Reply:
Added TODO [here](https://github.com/kubernetes/autoscaler/pull/5708/files#diff-03046c2a86c17ed94a0d62f14449afb658d5a297113cbf894f3a1150858d0203R122-R123)